src/Microsoft.Performance.SDK.Runtime/Discovery/AssemblyExtensionDiscovery.cs

```diff
@@ -203,70 +238,177 @@ public bool ProcessAssemblies(IEnumerable<string> directoryPaths)
+                            else
                             {
+                                assemblyErrors.Add(
+                                    new ErrorInfo(ErrorCodes.AssemblyLoadFailed, "The assembly failed to load.")
+                                    {
+                                        Target = filePath,
+                                    });
                                 allLoaded = false;
                             }
                         }
                     }
+
+                    if (assemblyErrors.Count > 0)
+                    {
+                        directoryErrors.Add(
+                            new ErrorInfo(ErrorCodes.PluginRegistrationFailed, "Unable to load the plugin.")
+                            {
+                                Target = directoryPath,
+                                Details = assemblyErrors.ToArray(),
+                            });
+                    }
                 }
 
                 Parallel.ForEach(this.observers, (observer) => observer.DiscoveryComplete());
             }
 
             watch.Stop();
             Console.Error.WriteLine("Loaded {0} in {1}ms", allLoaded ? "all" : "some", watch.ElapsedMilliseconds);
+            
+            if (directoryErrors.Count > 0)
+            {
+                Debug.Assert(!allLoaded);
+                error = new ErrorInfo(ErrorCodes.DiscoveryFailed, "One or more directories failed during the Plugin Discovery phase.")
+                {
+                    Target = string.Empty,
+                    Details = directoryErrors.ToArray(),
+                    Inner = new DiscoveryInnerError(
+                        directoryPaths,
+                        includeSubdirectories,
+                        searchPatterns,
+                        exclusionFileNames,
+                        exclusionsAreCaseSensitive),
+                };
+            }
+            else
+            {
+                Debug.Assert(allLoaded);
+                error = null;
+            }
+            
             return allLoaded;
         }
 
-        private bool ProcessAssembly(Assembly assembly)
+        private bool ProcessAssembly(
+            Assembly assembly,
+            out ErrorInfo error)
         {
             Guard.NotNull(assembly, nameof(assembly));
 
             var assemblyName = assembly.GetName().FullName;
-            bool success = false;
 
-            try
+            var referencedSdk = this.versionChecker.FindReferencedSdkVersion(assembly);
+            if (referencedSdk != null &&
+                !this.versionChecker.IsVersionSupported(referencedSdk))
             {
-                foreach (var type in assembly.GetTypes())
+                error = new ErrorInfo(
+                    ErrorCodes.SdkVersionIncompatible,
+                    "The assembly references a version of the SDK which is incompatible " +
+                    "with the hosting SDK. Please contact the author of the plugin.")
```

Comment:
I would remove "Please contact the author of the plugin," since this string should really just be details of what went wrong